### ISSUE TITLE
wgpu 0.19 regression spawning 160k entities with unique materials

### ISSUE THREAD

**Issue by superdump** [bevyengine/bevy#11946]
## Bevy version

`main` at `961d49f906`

## Relevant system information

Rust version: 1.76.0
macOS 14.3.1

`AdapterInfo { name: "Apple M1 Max", vendor: 0, device: 0, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Metal }`

## What you did

`cargo run --release --example bevymark -- --benchmark --waves 160 --per-wave 1000 --mode mesh2d --vary-per-instance`

This spawns 160k entities.

## What went wrong

- what were you expecting?

To see bouncing birds after about 5s as in Bevy 0.12.1.

- what actually happened?

I don't see any birds except the one on my Bevy hoodie until 50s.

## Additional information

I bisected and found that the following commit is the culprit:
```
commit 35ac1b152ee1b2c7b2930f67d93be0db6f395b47
Author: Elabajaba <Elabajaba@users.noreply.github.com>
Date:   Fri Jan 26 13:14:21 2024 -0500

    Update to wgpu 0.19 and raw-window-handle 0.6 (#11280)
```

**Comment by superdump** [comment]
This command spawns 160 waves of 1000 mesh2d entities. Every entity has a unique material. Because of how AsBindGroup currently works, this means 160k uniform buffers are created, and 160k bind groups.

If I capture a Tracy trace, 47s are spent in prepare_materials_2d: https://github.com/bevyengine/bevy/blob/961d49f90651bb0a6fe4a75db9d42df928ba3b26/crates/bevy_sprite/src/mesh2d/material.rs#L563

**Comment by rparrett** [comment]
It seems like wgpu 0.19.2 fixed this. Can you confirm?

**Comment by superdump** [comment]
I updated to macOS 14.4, and wgpu 0.19.3, and this seems to have got even worse? Now it hasn't finished after a few minutes...

**Comment by rparrett** [comment]
Okay, so I can't actually reproduce this with any version of wgpu and bevy 0.13.0. If someone can confirm that, I guess we could remove this from the 0.13.1 milestone.

However, I am seeing the same hang as Rob with recent commits on `main`. I bisected to #11986 and tested it and the previous commit several times.

```
AdapterInfo { name: "Apple M1 Max", vendor: 0, device: 0, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Metal }
SystemInfo { os: "MacOS 14.2.1 ", kernel: "23.2.0", cpu: "Apple M1 Max", core_count: "10", memory: "64.0 GiB" }
```

|bevy|wgpu|time to birb|
|-|-|-|
| 0.12.1 | 0.17.2 | 3.85 |
| 0.13.0 | =0.19.0 | 3.87 |
| 0.13.0 | =0.19.1 | 3.39 |
| 0.13.0 | =0.19.2 | 3.45 |
| 0.13.0 | =0.19.3 | 3.50 |
| main pre #11986 | =0.19.3 | 3.50 |
| main post #11986 | =0.19.3 | ??? hangs |

**Comment by superdump** [comment]
I take those measurements to mean that this is no longer a wgpu regression and is now a bevy regression on main. Thanks for testing @rparrett !

**Comment by rparrett** [comment]
That makes sense. I can reproduce the ~50 second slowdown if I pin the whole wgpu ecosystem to versions used in 0.19.1.

**Comment by superdump** [comment]
Yes, please.

I just verified here too that wgpu-core =0.19.0, wgpu =0.19.1 reproduces this issue, and it is fixed with both =0.19.2.

**Comment by rparrett** [comment]
Fixed upstream
